rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+use crate::{
+    coordinator::MaskHash,
+    mask::{Mask, MaskedModel},
+};
+use bincode;
+use futures::ready;
+use hex;
+use rusoto_core::{Region, RusotoError};
+use rusoto_s3::{
+    CreateBucketError,
+    CreateBucketOutput,
+    CreateBucketRequest,
+    Delete,
+    DeleteObjectsError,
+    DeleteObjectsOutput,
+    DeleteObjectsRequest,
+    GetObjectError,
+    GetObjectOutput,
+    GetObjectRequest,
+    ListObjectsV2Error,
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    ObjectIdentifier,
+    PutObjectError,
+    PutObjectOutput,
+    PutObjectRequest,
+    S3Client,
+    StreamingBody,
+    S3,
+};
+use std::{
+    collections::HashMap,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::stream::Stream;
+
+use thiserror::Error;
+use tokio::io::AsyncReadExt;
+use uuid::Uuid;
+
+#[derive(Debug, Error)]
+pub enum StorageError {
+    #[error("S3Error")]
+    S3Error,
+}
+
+#[derive(Clone)]
+struct Buckets(HashMap<&'static str, &'static str>);
+
+impl Buckets {
+    fn new() -> Self {
+        let mut names = HashMap::new();
+        names.insert("masks", "masks");
+        names.insert("masked-models", "masked-models");
+        names.insert("global-model", "global-model");
+        names.insert("global-masked-model", "global-masked-model");
+        Self(names)
+    }
+
+    fn masks(&self) -> &'static str {
+        self.0.get("masks").unwrap()
+    }
+
+    fn masked_models(&self) -> &'static str {
+        self.0.get("masked-models").unwrap()
+    }
+
+    fn global_model(&self) -> &'static str {
+        self.0.get("global-model").unwrap()
+    }
+
+    fn global_masked_model(&self) -> &'static str {
+        self.0.get("global-masked-model").unwrap()
+    }
+
+    fn names(&self) -> impl Iterator<Item = &&str> {
+        self.0.values().into_iter()
+    }
+}
+
+#[derive(Clone)]
+pub struct S3Store {
+    s3_client: S3Client,
+    buckets: Buckets,
+}
+
+// API
+impl S3Store {
+    pub fn new(region: Region) -> Self {
+        Self {
+            s3_client: S3Client::new(region),
+            buckets: Buckets::new(),
+        }
+    }
+
+    /// Upload a [`Mask`]
+    pub async fn upload_mask(&self, mask_hash: &MaskHash, mask: &Mask) -> Result<(), StorageError> {
+        // mask_hash is stored in the mask_dict
+        let se_mask = bincode::serialize(mask).unwrap();
+        let key = hex::encode(&mask_hash);
```

Comment:
Question: must keys be ascii? Or is it just user friendliness?

Reply:
No it doesn't have to. The only requirement is that the key is a string.